___

<a href='https://www.udemy.com/user/joseportilla/'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Content Copyright by Pierian Data</em></center>

- ## Refer (Jupyter Notebook Basics.ipynb) for Basics!

# Numbers and more in Python!

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Throughout this course we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

 
 
Now let's start with some basic arithmetic.

### Basic Arithmetic

In [1]:
# Addition
2+1

3

In [2]:
# Subtraction
2-1

1

In [3]:
# Multiplication
2*2

4

In [4]:
# Division
3/2

1.5

In [5]:
# Floor Division
7//4

1

**Whoa! What just happened? Last time I checked, 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. 

**Floor Division**: The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

                for ex:
                      If the result is 1.75 after the division, then floor division (//) turncates the decimal(deletes the                           decimal point) without rounding, and the result came out is 1 (an integer) 

### Modulo or 'Mod' Operator

In [5]:
# Modulo
7 % 4

3

**Modulo means Remainder**

Modulo or 'Mod' Operator is denoted by %

Modulo or 'Mod' operator and Classic Division method are the same!

In Classic Division, Dividend = Divisor * Quotient + Remainder

**'The answer of Modulo or 'Mod' Operator is the remainder of Classic Division'**


In [7]:
# Powers
2**3

8

The two asterisk ** means RAISED TO THE POWER. 

Here, the input is given as 2 RAISED TO THE POWER 3, which means cube of 2, which is 8.

In [8]:
# Can also do roots this way
4**0.5

2.0

We can also take Square Roots, Cube Roots of different numbers.

- For Square Root: (Integer)^1/2 
- For Cube Root  : (Integer)^1/3 
- For Quad Root  : (Integer)^1/4 

In [9]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In the above output, the answer is 105 rather than 156 because of BODMAS Rule.

In [10]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

In the above input of code, we can use parentheses to specify orders that which operation should take place first.